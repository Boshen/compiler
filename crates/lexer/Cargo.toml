[package]
name = "lexer"
version = "0.0.0"
edition = "2021"

[dependencies]
unicode-id = "0.3.0"

[dev-dependencies]
insta = "1.12.0"
unindent = "0.1.7"
