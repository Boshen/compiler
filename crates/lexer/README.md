# Lexer

## Goal

### Performance
Performance should be achieved by reducing the total amount of if checks in any code path.

## Fuzzing

```bash
cd crates/lexer && cargo +nightly fuzz run lexer -- -only_ascii=1
```
