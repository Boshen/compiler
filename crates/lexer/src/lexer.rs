//! Lexer

use crate::kind::Kind;
use crate::token::Token;

pub struct Lexer<'a> {
    /// The input string
    bytes: &'a [u8],

    /// The current index position
    cur: usize,

    /// Are we at the EOF?
    eof: bool,
}

impl<'a> Lexer<'a> {
    #[must_use]
    pub const fn new(input: &'a str) -> Self {
        Self {
            bytes: input.as_bytes(),
            cur: 0,
            eof: false,
        }
    }
}

impl Iterator for Lexer<'_> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        if self.cur >= self.bytes.len() {
            if self.eof {
                return None;
            }
            self.eof = true;
            return Some(Token::new(Kind::EOF, self.cur, 1));
        }

        let token = self
            .read_whitespace()
            .or_else(|| self.read_identifier())
            .or_else(|| Some(Token::new(Kind::Unknown, self.cur, 1)));

        self.cur += token.as_ref().unwrap().len();

        token
    }
}

impl Lexer<'_> {
    fn read_whitespace(&mut self) -> Option<Token> {
        let mut len = 0;

        for chr in &self.bytes[self.cur..] {
            if !chr.is_ascii_whitespace() {
                break;
            }
            len += 1;
        }

        if len == 0 {
            return None;
        }

        Some(Token::new(Kind::WhiteSpace, self.cur, len))
    }

    fn read_identifier(&mut self) -> Option<Token> {
        let mut len = 0;

        for chr in &self.bytes[self.cur..] {
            if !chr.is_ascii_alphabetic() {
                break;
            }
            len += 1;
        }

        if len == 0 {
            return None;
        }

        let kind = match &self.bytes[self.cur..self.cur + len] {
            b"null" => Kind::Null,
            b"undefined" => Kind::Undefined,
            b"true" => Kind::True,
            b"false" => Kind::False,
            _ => Kind::Ident,
        };

        Some(Token::new(kind, self.cur, len))
    }
}
