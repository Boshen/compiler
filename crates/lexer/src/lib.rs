#![warn(clippy::all, clippy::pedantic, clippy::nursery)]

mod constants;
mod kind;
mod lexer;
mod state;
mod token;

pub use crate::kind::*;
pub use crate::lexer::*;
pub use crate::token::*;
