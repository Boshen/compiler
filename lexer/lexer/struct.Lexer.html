<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=../../lexer/struct.Lexer.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="../../lexer/struct.Lexer.html">../../lexer/struct.Lexer.html</a>...</p>
    <script>location.replace("../../lexer/struct.Lexer.html" + location.search + location.hash);</script>
</body>
</html>