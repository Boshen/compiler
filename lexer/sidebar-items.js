initSidebarItems({"enum":[["Kind",""],["Number",""]],"struct":[["Lexer",""],["Token",""]]});