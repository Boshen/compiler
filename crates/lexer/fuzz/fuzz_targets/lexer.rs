#![no_main]

use lexer::Lexer;
use libfuzzer_sys::fuzz_target;

fuzz_target!(|data: &[u8]| {
    if let Ok(s) = std::str::from_utf8(data) {
        Lexer::new(s).into_iter().count();
    }
});
