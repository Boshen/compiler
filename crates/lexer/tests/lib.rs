#![warn(clippy::all, clippy::pedantic, clippy::nursery)]
#[cfg(test)]
use lexer::Lexer;
use unindent::unindent;

fn test_snapshot(name: &str, input: &str) {
    let input = unindent(input);
    let tokens = Lexer::new(input.trim())
        .into_iter()
        .filter(|t| !t.kind().is_whitespace())
        .collect::<Vec<_>>();
    let snapshot = format!("# Input\n{}\n---\n# Output\n{:#?}", input, tokens);
    insta::with_settings!({
        prepend_module_to_snapshot => false,
    }, {
        insta::assert_snapshot!(name, snapshot, name);
    });
}

#[test]
fn comment() {
    let input = "
        // comment
        /* multiline comment */
    ";
    test_snapshot("comment", input);
}

#[test]
fn keyword() {
    let input = "
        await break case catch class const continue debugger default delete do else enum export extends
        false finally for function if import in instanceof new null return super switch this throw true try typeof var void while with yield
        undefined
    ";
    test_snapshot("keyword", input);
}

#[test]
fn identifier() {
    let input = r#"
        $ _ $a _a abc_$
        \u03bc
        \u{61}
    "#;
    test_snapshot("identifier", input);
}

#[test]
fn punctuator() {
    let input = "
        { ( ) [ ] . ... ; , < > <= >= == != === !== + - * % ** ++ -- << >> >>> & | ^ ! ~ && || ?? ? : = +=
        -= *= %= **= <<= >>= >>>= &= |= ^= &&= ||= ??= =>
        ?. / /= }
        #
    ";
    test_snapshot("punctuator", input);
}

#[test]
fn numeric_literal() {
    let input = "
        0 0n
        0b1 0B12
        0o1 0O12
        0x1 0X12
        0123 0789
        0. 0.0 0.123
        123n
        1.0 1.1
    ";
    test_snapshot("numeric_literal", input);
}

#[test]
fn string_literal() {
    let input = r#"
        "12345" '12345'
        "" ''
        '\\\n\r\t\b\v\f\'\"'
        '\\\n\r\t\b\v\f\'\"'
        '\u1234' '\x12'
        'foo \
        '
        "\d"
    "#;
    test_snapshot("string_literal", input);
}

#[test]
fn regex() {
    let input = r#"
        /aa/;
        /[0-9A-Za-z_\$(|)\[\]\/\\^]/;
        /[//]/;
        /[/]/;
    "#;
    test_snapshot("regex", input);
}

#[test]
fn template_literal() {
    let input = r#"
        ``
        `123`
        `\`\r`
    "#;
    test_snapshot("template_literal", input);
}
