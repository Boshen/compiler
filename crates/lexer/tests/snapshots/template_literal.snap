---
source: crates/lexer/tests/lib.rs
assertion_line: 16
expression: template_literal

---
# Input
``
`123`
`\`\r`

---
# Output
[
    Token {
        kind: Template,
        offset: 0,
        len: 2,
    },
    Token {
        kind: Template,
        offset: 3,
        len: 5,
    },
    Token {
        kind: Template,
        offset: 9,
        len: 6,
    },
    Token {
        kind: EOF,
        offset: 15,
        len: 1,
    },
]
